examples/vision/captcha_ocr.py

```diff
@@ -0,0 +1,334 @@
+from tensorflow.keras import layers
+
+seed = 1234
+np.random.seed(seed)
+tf.random.set_seed(seed)
+
+"""
+## Load the data: [Captcha Images](https://www.kaggle.com/fournierp/captcha-version-2-images)
+Let's download the data.
+"""
+
+
+"""shell
+curl -LO https://github.com/AakashKumarNain/CaptchaCracker/raw/master/captcha_images.tar.gz
+tar -zxf captcha_images.tar.gz
+"""
+
+
+"""
+The dataset contains 1070 captcha files as png images. The label for each sample is the
+name of the file (excluding the '.png' part). We will load the data as numpy arrays. The label
+for each sample is a string. We will map each character in the string to a number for training
+the model. Similary, we would be required to map the predictions of the model back to string.
+For this purpose would maintain two dictionary mapping characters to numbers and numbers to
+characters respectively.
+"""
+
+
+# Path to the data directory
+data_dir = Path("./captcha_images/samples/")
+
+# Get list of all the images
+images = sorted(list(map(str, list(data_dir.glob("*.png")))))
+labels = [img.split(os.path.sep)[-1].split(".png")[0] for img in images]
+characters = set()
+
+
+for label in labels:
+    for char in label:
+        characters.add(char)
+
+print("Number of images found: ", len(images))
+print("Number of labels found: ", len(labels))
+print("Number of unique characters: ", len(characters))
+print("Characters present: ", characters)
+
+# Batch size for training and validation
+batch_size = 16
+
+# Desired image dimensions
+img_width = 200
+img_height = 50
+
+# Factor  by which the image is going to be downsampled
+# by the convolutional blocks. We will be using two
+# convolution blocks and each convolution block will have
+# a pooling layer which downsample the features by a factor of 2.
+# Hence total downsampling factor would be 4.
+downsample_factor = 4
+
+# Maximum length of any captcha in the dataset
+max_length = 5
+
+
+"""
+## Preprocessing
+"""
+
+
+## Mapping characters to numbers
+char_to_num = layers.experimental.preprocessing.StringLookup(
+    vocabulary=list(characters), num_oov_indices=0, mask_token=None
+)
+
+# Mapping numbers back to original characters
+num_to_char = layers.experimental.preprocessing.StringLookup(
+    vocabulary=char_to_num.get_vocabulary(), mask_token=None, invert=True
+)
+
+# Splitting data into training and validation sets
+X_train, X_valid, y_train, y_valid = train_test_split(
+    np.array(images), np.array(labels), test_size=0.1, random_state=seed, shuffle=True
+)
+
+
+def encode_single_sample(img_path, label):
+    # 1. Read image
+    img = tf.io.read_file(img_path)
+    img = tf.io.decode_png(img, channels=3)
+    # 2. Convert to grayscale
+    img = tf.image.rgb_to_grayscale(img)
+    # 3. Convert to float32 in [0, 1] range
+    img = tf.image.convert_image_dtype(img, tf.float32)
+    # 4. Resize to the desired size
+    img = tf.image.resize(img, [img_height, img_width])
+    # 5. Transpose the image
```

Comment:
Why?

Reply:
Valid question. The images are in this dataset have a shape `(50, 200, 3)`. If we don't transpose, then after  the two conv blocks and reshape, we would have a shape of `(12, 3200)` (as compared to `(50, 768)` when using transposed version). The number of features that the RNN can attend to is very small and the network fails to converge/perform very poorly.  Please let me know if you want me to add this as a comment

Reply:
Yes, please add a comment -- otherwise it's non-obvious. Basically you want the time dimension to correspond to the width of the image